### ISSUE TITLE
Cannot import name 'AsyncMultiKernelManager' with custom notebook image

### ISSUE THREAD

**Issue by Ursanon** [kubeflow/kubeflow#5665]
I've added new packages (`ffmpeg, libsm6, libxext6`) to existing tf notebook dockerfile.
Pod is affected by CrashLoopBackOff, log:

```
Traceback (most recent call last):
  File "/usr/local/bin/jupyter-lab", line 5, in <module>
    from jupyterlab.labapp import main
  File "/usr/local/lib/python3.6/dist-packages/jupyterlab/__init__.py", line 7, in <module>
    from .labapp import LabApp
  File "/usr/local/lib/python3.6/dist-packages/jupyterlab/labapp.py", line 15, in <module>
    from jupyterlab_server import slugify, WORKSPACE_EXTENSION
  File "/usr/local/lib/python3.6/dist-packages/jupyterlab_server/__init__.py", line 4, in <module>
    from .app import LabServerApp
  File "/usr/local/lib/python3.6/dist-packages/jupyterlab_server/app.py", line 11, in <module>
    from jupyter_server.extension.application import ExtensionApp, ExtensionAppJinjaMixin
  File "/usr/local/lib/python3.6/dist-packages/jupyter_server/extension/application.py", line 21, in <module>
    from jupyter_server.serverapp import ServerApp
  File "/usr/local/lib/python3.6/dist-packages/jupyter_server/serverapp.py", line 75, in <module>
    from .services.kernels.kernelmanager import MappingKernelManager, AsyncMappingKernelManager
  File "/usr/local/lib/python3.6/dist-packages/jupyter_server/services/kernels/kernelmanager.py", line 20, in <module>
    from jupyter_client.multikernelmanager import MultiKernelManager, AsyncMultiKernelManager
ImportError: cannot import name 'AsyncMultiKernelManager'
```

**Comment by DavidSpek** [comment]
@Ursanon Which Dockerfile are you using exactly?

**Comment by Ursanon** [comment]
@DavidSpek [This one](https://github.com/kubeflow/kubeflow/tree/master/components/tensorflow-notebook-image): `components/tensorflow-notebook-image`

**Comment by DavidSpek** [comment]
@Ursanon As you might have noticed that Dockerfile has not been updated in a long time and I wouldn't recommend using it. I have been working on creating new Dockerfiles and images for Kubeflow, and while these images are not yet final (and hosted on my private Dockerhub), maybe you can try with the following Dockerfiles as an example. 
https://github.com/kubeflow/kubeflow/pull/5632/files

If you require any more help or would like a Dockerfile a completely custom Dockerfile that starts from `ubuntu` and gives you tensorflow and the packages you need I can create that for you tomorrow if you tell me what packages you need installed.

**Comment by Ursanon** [comment]
I've used tensorflow 2.2.0 jupyter notebook base image, but it gave me:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/traitlets/traitlets.py", line 528, in get
    value = obj._trait_values[self.name]
KeyError: 'runtime_dir'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/jupyter-notebook", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/dist-packages/jupyter_core/application.py", line 268, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/traitlets/config/application.py", line 663, in launch_instance
    app.initialize(argv)
  File "</usr/local/lib/python3.6/dist-packages/decorator.py:decorator-gen-7>", line 2, in initialize
  File "/usr/local/lib/python3.6/dist-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/notebook/notebookapp.py", line 1717, in initialize
    self.init_configurables()
  File "/usr/local/lib/python3.6/dist-packages/notebook/notebookapp.py", line 1372, in init_configurables
    connection_dir=self.runtime_dir,
  File "/usr/local/lib/python3.6/dist-packages/traitlets/traitlets.py", line 556, in __get__
    return self.get(obj, cls)
  File "/usr/local/lib/python3.6/dist-packages/traitlets/traitlets.py", line 535, in get
    value = self._validate(obj, dynamic_default())
  File "/usr/local/lib/python3.6/dist-packages/jupyter_core/application.py", line 99, in _runtime_dir_default
    ensure_dir_exists(rd, mode=0o700)
  File "/usr/local/lib/python3.6/dist-packages/jupyter_core/utils/__init__.py", line 13, in ensure_dir_exists
    os.makedirs(path, mode=mode)
  File "/usr/lib/python3.6/os.py", line 210, in makedirs
    makedirs(head, mode, exist_ok)
  File "/usr/lib/python3.6/os.py", line 210, in makedirs
    makedirs(head, mode, exist_ok)
  File "/usr/lib/python3.6/os.py", line 210, in makedirs
    makedirs(head, mode, exist_ok)
  File "/usr/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/home/jovyan/.local'
```
every volume mounted by kubeflow is assigned to `root` user.

**Comment by DavidSpek** [comment]
@Ursanon Can you tell me what packages you need (and if you need tensorflow GPU support)? Then I will create a Dockerfile for you tomorrow so that you can continue your work while the new images are still being reviewed.

**Comment by Ursanon** [comment]
Thank you so much for your help @DavidSpek !

I need opencv and torch (+ tensorboard with tf2) and yes, I need GPU support.

**Comment by DavidSpek** [comment]
@Ursanon Just to be clear, do you need tensorflow or PyTorch?

**Comment by Ursanon** [comment]
I need both: opencv + tf2 and opencv + pytorch

**Comment by DavidSpek** [comment]
Both in a single image, or 2 separate ones? If possible I'd go for 2 separate ones to avoid 2 versions of cuda being installed in a single image which will make the image very large.

**Comment by Ursanon** [comment]
Yeah, two separate images seems more reasonable.

**Comment by DavidSpek** [comment]
@Ursanon I've been a bit busier than expected, so I haven't extensively tested the images but they should work. If something doesn't work, please let me know and I will fix it as the plan is to use these images for the upcoming Kubeflow release in 9 days.

```
FROM davidspek/kubeflow-ubuntu-pytorch-base:0.3-cuda

RUN python3 -m pip install --quiet --no-cache-dir\
    tensorboard==2.4.1 \
    opencv-python==4.5.1.48 
```

```
FROM davidspek/kubeflow-ubuntu-tensorflow-base:0.2-cuda

RUN python3 -m pip install --quiet --no-cache-dir\
    tensorboard==2.4.1 \
    opencv-python==4.5.1.48 
```

The base images they use are from https://github.com/kubeflow/kubeflow/pull/5626 and https://github.com/kubeflow/kubeflow/pull/5628.

You can grab the image from `davidspek/ursanon-kubeflow:0.1-pytorch` and `davidspek/ursanon-kubeflow:0.1-tensorflow`.

**Comment by Ursanon** [comment]
@DavidSpek Thanks for your work.

Is base image open source?

Unfortunately opencv requires some libs: `ImportError: libGL.so.1: cannot open shared object file: No such file or directory`

**Comment by DavidSpek** [comment]
@Ursanon The base images are all open source and start from Ubuntu. I linked the PR's they came from, but they will be merged into the repo and official images will be out soon. 

You can try building these and see if that helps. 
```
FROM davidspek/kubeflow-ubuntu-pytorch-base:0.3-cuda

USER root
RUN apt-get -yq update \
         && apt-get -yq install --no-install-recommends \
         libsm6 \
         libxext6 \
         && apt-get clean \
         && rm -rf /var/lib/apt/lists/*
USER $NB_UID
RUN python3 -m pip install --quiet --no-cache-dir\
    tensorboard==2.4.1 \
    opencv-python==4.5.1.48
```

```
FROM davidspek/kubeflow-ubuntu-tensorflow-base:0.2-cuda

USER root
RUN apt-get -yq update \
         && apt-get -yq install --no-install-recommends \
         libsm6 \
         libxext6 \
         && apt-get clean \
         && rm -rf /var/lib/apt/lists/*
USER $NB_UID
RUN python3 -m pip install --quiet --no-cache-dir\
    tensorboard==2.4.1 \
    opencv-python==4.5.1.48
````

**Comment by jbottum** [comment]
@Ursanon Hi, thanks for your request.    I hope that you can support your requirements via Community support options.   If you get into a time crunch or need enterprise level support, you might consider asking your platform team (AWS, GCP, Azure) or a vendor that provides support (Arrikto, etc).

**Comment by Ursanon** [comment]
@DavidSpek Thanks for base images. 
I've modified them to fit my needs.

**Comment by DavidSpek** [comment]
@Ursanon Do they work as expected? Are there any other issues you are experiencing? In the next few days the Kubeflow images should be up so you can then migrate to those rather than using the ones from my personal dockerhub.

**Comment by Ursanon** [comment]
So far so good, have some more tests to process. I'll keep you posted.

**Comment by Ursanon** [comment]
@DavidSpek I encountered an error while finetuning instance segmentation demo from pytorch tutorial.

Error:
```
/opt/conda/lib/python3.8/site-packages/torchvision/models/detection/_utils.py in __call__(self, matched_idxs)
     55             # randomly select positive and negative examples
     56             perm1 = torch.randperm(positive.numel(), device=positive.device)[:num_pos]
---> 57             perm2 = torch.randperm(negative.numel(), device=negative.device)[:num_neg]
     58 
     59             pos_idx_per_image = positive[perm1]

RuntimeError: radix_sort: failed on 1st step: cudaErrorInvalidDevice: invalid device ordinal
```

`nvidia-smi` output:
```
Thu Mar 11 12:11:53 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P40           On   | 00000001:00:00.0 Off |                    0 |
| N/A   18C    P8    11W / 250W |      0MiB / 22919MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

**Comment by DavidSpek** [comment]
@Ursanon Thank you for reporting the issue. Just so I am clear on the actual problem, was this a specific step that encountered an error but were you able to perform other tasks (for example training) successfully on the GPU using the image? Or were you not able to perform any tasks on the GPU with your image?

**Comment by Ursanon** [comment]
@DavidSpek sorry for late answer, I wasn't able to train my model using GPU.

**Comment by DavidSpek** [comment]
@Ursanon No problem, I have have also been very busy. Maybe you can try to swap out the base image for the one listed in https://github.com/kubeflow/kubeflow/pull/5626. I just tested the CUDA one from https://github.com/kubeflow/kubeflow/pull/5631 (which uses the previous one as a base) and I can train on my RTX 3090 just fine. 
```
Mon Mar 15 09:54:36 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.39       Driver Version: 460.39       CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce RTX 3090    Off  | 00000000:42:00.0 Off |                  N/A |
| 48%   55C    P2   121W / 350W |   2267MiB / 24268MiB |      7%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

**Comment by DavidSpek** [comment]
@Ursanon Looking at the error you posted it seems like you might be trying to run the code on the wrong device. 
https://stackoverflow.com/questions/22175825/cuda-invalid-device-ordinal
https://www.programmersought.com/article/71025630639/

**Comment by Ursanon** [comment]
I have single gpu and moved everything to `device = torch.device('cuda')`.
Operations on tensors were ok, just training on torchvision's sample threw this error.

Will investigate this issue further.

**Comment by DavidSpek** [comment]
@Ursanon Maybe you can try the following code that runs a MNIST example on the GPU:

```
from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout(0.25)
        self.dropout2 = nn.Dropout(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output

def train(model, device, train_loader, optimizer, epoch, log_interval, dry_run):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))
            if dry_run:
                break

def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

def main(train_batch_size, test_batch_size, epochs, lr, gamma, no_cuda, dry_run, seed, log_interval, save_model):
    use_cuda = not no_cuda #and torch.cuda.is_available()

    torch.manual_seed(seed)

    device = torch.device("cuda" if use_cuda else "cpu")

    train_kwargs = {'batch_size': train_batch_size}
    test_kwargs = {'batch_size': test_batch_size}
    if use_cuda:
        cuda_kwargs = {'num_workers': 1,
                       'pin_memory': True,
                       'shuffle': True}
        train_kwargs.update(cuda_kwargs)
        test_kwargs.update(cuda_kwargs)

    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset1 = datasets.MNIST('./data', train=True, download=True,
                       transform=transform)
    dataset2 = datasets.MNIST('./data', train=False,
                       transform=transform)
    train_loader = torch.utils.data.DataLoader(dataset1,**train_kwargs)
    test_loader = torch.utils.data.DataLoader(dataset2, **test_kwargs)

    model = Net().to(device)
    optimizer = optim.Adadelta(model.parameters(), lr=lr)

    scheduler = StepLR(optimizer, step_size=1, gamma=gamma)
    for epoch in range(1, epochs + 1):
        train(model, device, train_loader, optimizer, epoch, log_interval, dry_run)
        test(model, device, test_loader)
        scheduler.step()

    if save_model:
        torch.save(model.state_dict(), "mnist_cnn.pt")

main(train_batch_size=64,
     test_batch_size=1000,
     epochs=2, lr=1.0,
     gamma=0.7,
     no_cuda=False,
     dry_run=False,
     seed=1,
     log_interval=10,
     save_model=False)
```

**Comment by Ursanon** [comment]
@DavidSpek Unfortunately provided example returns `server error 503`.

[This issue](https://github.com/kubeflow/kubeflow/issues/5665#issuecomment-796708138) was raised when testing with provided jupyter notebook: [MaskRCNN-testTrain.zip](https://github.com/kubeflow/kubeflow/files/6155684/MaskRCNN-testTrain.zip)

**Comment by DavidSpek** [comment]
@Ursanon When are you seeing `server error 503` exactly?

**Comment by Ursanon** [comment]
@DavidSpek Try visiting http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz

**Comment by DavidSpek** [comment]
@Ursanon you can grab the 4 zips from https://github.com/golbin/TensorFlow-MNIST/tree/master/mnist/data as well.

**Comment by Ursanon** [comment]
@DavidSpek I found that every training with `troch==1.8.0` fails with the same error (`torch radix_sort: failed on 1st step: cudaErrorInvalidDevice: invalid device ordinal`) even on different GPUs:

After downgrade to `torch==1.7.1` everything works fine

**Comment by DavidSpek** [comment]
@Ursanon Thanks for confirming that. The image that is being included with Kubeflow 1.3 will be using torch 1.7.1 as it stands now. We'll need to investigate what is causing this with torch 1.8 as I wasn't experiencing this issue personally.

**Comment by Ursanon** [comment]
folks on pyTorch discuss have the same problem with torch 1.8
https://discuss.pytorch.org/t/pytorch-1-8-0-fasterrcnn-resnet50-fpn-error/114150/4
